lib/rules/media-query-no-invalid/index.mjs

```diff
@@ -59,9 +71,25 @@ const rule = (primary) => {
 				}
 
 				mediaQuery.walk(({ node, parent }) => {
-					// All general enclosed nodes are invalid.
 					if (isGeneralEnclosed(node)) {
-						complain(atRule, atRuleParamIndexValue, node);
+						let shouldIgnore = true;
+						let hasFunction = false;
+
+						node.walk(({ node: childNode }) => {
+							if (isFunctionNode(childNode)) {
+								hasFunction = true;
+								const functionName = childNode.getName();
+
+								if (!optionsMatches(secondaryOptions, 'ignoreFunctions', functionName)) {
+									shouldIgnore = false;
+								}
+							}
+						});
+
+						// Do not report if all custom functions are ignored.
+						if (!hasFunction || !shouldIgnore) {
+							complain(atRule, atRuleParamIndexValue, node);
+						}
```

Comment:
I think it might be better to ignore the entire `GeneralEnclosed` if it contains anything that is ignored. (currently we only ignore functions)

This allows for a simpler implementation and is more in line with how we handle non-standard syntax in general. (any non-standard function will always become `GeneralEnclosed`)

I also don't think there is any way we can partially ignore contents of a `GeneralEnclosed`. The only option to have such behavior would be by replacing the non-standard syntax with a standard equivalent and to then re-evaluate.

```js
if (isGeneralEnclosed(node)) {
	let ignored = false;

	if ('ignoreFunctions' in secondaryOptions) { // no need to walk if there are no ignore options
		node.walk(({ node: childNode }) => {
			if (!isFunctionNode(childNode)) {
				return;
			}

			ignored = optionsMatches(secondaryOptions, 'ignoreFunctions', childNode.getName());

			if (ignored) {
				return false; // returning `false` stops iterations
			}
		});
	}

	if (!ignored) {
		complain(atRule, atRuleParamIndexValue, node);
	}

	return;
}
```

Reply:
```css
@media (min-width: var(300px)), (min-width: theme(screens.md)) {}
```

Comma separated media query lists should still report errors

Reply:
@romainmenke 
Let me confirm your suggestion.

```css
@media (min-width: theme(screens.md) && fooFunction(300px)) {}
```

If `theme` is in the `ignoreList`:

- Currently, `theme` is ignored, but since `fooFunction` is not, an error is triggered.
- In your proposal, because `theme` is in the ignore list, the entire `GeneralEnclosed` block (including `fooFunction`) would be ignored, so no error would be thrown.

Is this understanding correct?

Reply:
Yes, that is correct.

There simply is no way to recover after ignoring a part of a `GeneralEnclosed`.

Some samples:
- `@media (min-width: calc(theme(screens.md) + 10px) {}` -> no error?
- `@media (min-width: calc(theme(screens.md) + 10not-a-px) {}` -> an error?
- `@media (min-width: theme(screens.md fooFunction()) {}` -> no error?

Because we don't know what the ignored function is supposed to be we can only assume that the entire `GeneralEnclosed` becomes valid and standard CSS once the ignored function is de-sugared.

Reply:
Ah, I understand. Thank you for the clear explanation!

Reply:
@romainmenke 

> `@media (min-width: var(300px)), (min-width: theme(screens.md)) {}`
Comma separated media query lists should still report errors

Since commas split media queries, each part can be validated independently, and errors should be triggered for any invalid part. So, doesn’t the current behavior seem correct?

Or, similar to how we handle `GeneralEnclosed`, should the entire media query list be skipped if one part contains an ignored function?

Reply:
Yeah, it's just an extra test to ensure that we are not ignoring too much if a single `GeneralEnclosed` is ignored :)

Current implementation/behavior should already be correct for that.
Just a bit of extra test coverage.